pallets/runtime/src/bridge.rs

```diff
@@ -232,11 +308,65 @@ decl_module! {
                     bridge_tx
                 }));
             } else {
-                <HandledProposals<T>>::insert(&bridge_tx, true);
+                <HandledTxs<T>>::insert(&bridge_tx, true);
                 Self::deposit_event(RawEvent::Bridged(bridge_tx));
             }
             Ok(())
         }
+
+        /// Freezes given bridge transactions.
+        pub fn freeze_txs(origin, bridge_txs: Vec<BridgeTx<T::AccountId, T::Balance>>) ->
+            DispatchResult
+        {
+            Self::check_admin(origin)?;
+            for bridge_tx in bridge_txs {
+                let proposal =
+                    <T as Trait>::Proposal::from(Call::<T>::handle_bridge_tx(bridge_tx.clone())).into();
+                let proposal_id = <multisig::Module<T>>::proposal_ids(&Self::relayers(), &proposal);
+                ensure!(proposal_id.is_some(), Error::<T>::NoSuchProposal);
+                <FrozenTxs<T>>::insert(&bridge_tx, true);
+                Self::deposit_event(RawEvent::FrozenTx(bridge_tx));
+            }
+            Ok(())
+        }
+
+        /// Unfreezes given bridge transactions.
+        pub fn unfreeze_txs(origin, bridge_txs: Vec<BridgeTx<T::AccountId, T::Balance>>) ->
+            DispatchResult
+        {
+            Self::check_admin(origin)?;
+            for bridge_tx in bridge_txs {
+                ensure!(Self::frozen_txs(&bridge_tx), Error::<T>::NoSuchFrozenTx);
+                <FrozenTxs<T>>::remove(&bridge_tx);
+                Self::deposit_event(RawEvent::UnfrozenTx(bridge_tx.clone()));
+                ensure!(!Self::handled_txs(&bridge_tx), Error::<T>::ProposalAlreadyHandled);
```

Comment:
This check should be done before removing `bridge_tx` from `FrozenTxs`. Storage changes already done in a tx are not reverted by ensure. Unlikely that a tx will be in both `frozen_txs` and `handled_txs` but still :)

Reply:
The present ordering allows to remove the frozen status from already handled transactions.

It is possible that a transaction is both handled and frozen because the admin can freeze any transactions and no `handled_txs` check is performed then. An admin can freeze an already handled transaction by mistake or because of some off-chain latency.

Reply:
If it's possible to hard fail after a state change, then we'll need to fix the ordering as this invariant needs to be maintained.